src/natcap/invest/globio.py

```diff
@@ -939,22 +939,29 @@ def _collapse_infrastructure_layers(
 
     def _collapse_infrastructure_op(*infrastructure_array_list):
         """For each pixel, create mask 1 if all valid, else set to nodata."""
-        nodata_mask = (
-            numpy.isclose(
-                infrastructure_array_list[0], infrastructure_nodata_list[0]))
-        infrastructure_result = infrastructure_array_list[0] > 0
+        infrastructure_result = numpy.empty_like(
+            infrastructure_array_list[0], dtype=numpy.int32)
+        # Zero's indicate where there's no infrastructure influence but not
+        # necessarily nodata
+        infrastructure_result[:] = 0
+
+        nodata_mask = numpy.isclose(
+            infrastructure_array_list[0], infrastructure_nodata_list[0])
+
+        infrastructure_mask = infrastructure_array_list[0] > 0 & ~nodata_mask
+
         for index in range(1, len(infrastructure_array_list)):
             current_nodata = numpy.isclose(
                 infrastructure_array_list[index],
                 infrastructure_nodata_list[index])
 
-            infrastructure_result = (
-                infrastructure_result |
+            infrastructure_mask = (
+                infrastructure_mask |
                 ((infrastructure_array_list[index] > 0) & ~current_nodata))
 
-            nodata_mask = (
-                nodata_mask & current_nodata)
+            nodata_mask = (nodata_mask & current_nodata)
 
+        infrastructure_result[infrastructure_mask] = 1
```

Comment:
Can you also explain why `infrastructure_mask` is not the result and needs to be separated into `infrastructure_result`. If I'm reading this right, `infrastructure_result` will be 0 where `infrastructure_mask` is false and 1 where it is true. Is that not right?

Reply:
That is exactly right, but I don't think we want to actually return a boolean array here. We can't properly set nodata values into a boolean array because the type conflict will just set the boolean array to either 0 or 1 depending on a positive or negative nodata value. Currently the model sets that nodata value to -1. 

I also think it's weird to return a boolean array from a function that writes a raster with GDAL which doesn't have a boolean type, but should be a Byte type.

Reply:
Ahh got it, I missed that there is a valid nodata case there. I mentioned below I think the solution would be to initialize `infrastructure_result` to a `numpy.uint8` and choose a nodata value of 2 or 255 would handle this. (p.s. if you choose 1 it turns into 255 anyway but best not to bit twiddle).

Reply:
haha, no one wants to be a bit twiddle. I like it!